### PR TITLE
[INTEL MKL] Add MKL-DNN quantized Matmul op with some fusions - Part3.

### PR DESCRIPTION
**Pull request by mdfaijul**
This PR is to replace an older PR #26271 by splitting into three. Current PR is part3. It adds unit tests for MKL-DNN quantized Matmul ops. Tests depends on part1 (#26909) and part2 (#26910).

### CODE REVIEW

**Review by penpornk — COMMENTED** [review]
Just a note that we will look at this once we are done with its dependencies (parts 1 and 2).

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+#include "tensorflow/core/framework/types.pb.h"
+#include "tensorflow/core/kernels/ops_testutil.h"
+#include "tensorflow/core/kernels/ops_util.h"
+#include "tensorflow/core/kernels/quantization_utils.h"
+#include "tensorflow/core/lib/core/status_test_util.h"
+#include "tensorflow/core/platform/test.h"
+
+namespace tensorflow {
+
+// Helper class for converting MKL tensors to TF tensors and comparing to
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op_test.cc]: Please move this comment to directly above `ConvMklToTF`.

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+// expected values
+
+static const uint8 dummy_tensor[] = {0, 0, 0, 0, 0, 0, 0, 0};
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op_test.cc]: Per our [Style Guide](https://google.github.io/styleguide/cppguide.html#Constant_Names) for constant variables,
```suggestion
static const uint8 kDummyTensor[] = {0, 0, 0, 0, 0, 0, 0, 0};
```

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+static const TensorShape dummy_shape({2, 4});
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op_test.cc]:
```suggestion
static const TensorShape kDummyShape({2, 4});
```

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+  TF_ASSERT_OK(
+      NodeDefBuilder("quantized_mat_mul_op", "_MklQuantizedMatMulWithBias")
+          .Input(FakeInput(DT_QUINT8))
+          .Input(FakeInput(DT_QINT8))
+          .Input(FakeInput(DT_QINT32))
+          .Input(FakeInput(DT_FLOAT))
+          .Input(FakeInput(DT_FLOAT))
+          .Input(FakeInput(DT_FLOAT))
+          .Input(FakeInput(DT_FLOAT))
+          .Input(FakeInput(DT_UINT8))  // MKl second tensor
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op_test.cc]:
```suggestion
          .Input(FakeInput(DT_UINT8))  // MKL second tensor
```
Please fix other places too.

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+  // Here are the results we expect, from hand calculations:
+  // (1 * 7) + (2 * 11) + (3 * 15) = 74
+  // (1 * 8) + (2 * 12) + (3 * 16) = 80
+  // (1 * 9) + (2 * 13) + (3 * 17) = 86
+  // (1 * 10) + (2 * 14) + (3 * 18) = 92
+  // (4 * 7) + (5 * 11) + (6 * 15) = 173
+  // (4 * 8) + (5 * 12) + (6 * 16) = 188
+  // (4 * 9) + (5 * 13) + (6 * 17) = 203
+  // (4 * 10) + (5 * 14) + (6 * 18) = 218
+  // final result after Bias addition
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op_test.cc]:
```suggestion
  // Final result after Bias addition:
```

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op_test.cc]:
```suggestion
  // Final result after Bias addition:
```

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc [hunk 7]
```diff
@@ -0,0 +1,464 @@
...
+  const Tensor& mkl_shape_tensor = *GetOutput(3);
+  ConvMklToTF conv_comp;
+  Tensor output_quantized;
+  conv_comp.ConvertMKL2TF<qint32>(DT_QINT32, output, mkl_shape_tensor,
+                                  output_quantized);
+
+  test::ExpectTensorEqual<qint32>(expected, output_quantized);
+}
+
+// Two small matrices A of type uint8(converted from signed integer)
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op_test.cc]:
```suggestion
// Two small matrices A of type uint8 (converted from signed integer)
```

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc [hunk 8]
```diff
@@ -0,0 +1,464 @@
...
+          .Attr("_kernel", "QuantizedMklOp")
+          .Finalize(node_def()));
+  TF_ASSERT_OK(InitOp());
+  // The A matrix is:
+  // |  -1 |  -5 |  -9 |
+  // |  -2 |  -6 | -10 |
+  // |  -3 |  -7 | -11 |
+  // |  -4 |  -8 | -12 |
+  // The input array only contains unsigned bytes, so we specify the actual
+  // quantized values as Au8 = (Af32 - Min(A𝑓32)*Qa, where
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op_test.cc]:
```suggestion
  // quantized values as Au8 = (Af32 - Min(A𝑓32)) * Qa, where
```

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc [hunk 9]
```diff
@@ -0,0 +1,464 @@
...
+  AddInputFromArray<float>(TensorShape({1}), {127.0f});
+  AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
+  AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
+  AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
+  AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
+  AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
+  AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
+  AddInputFromArray<uint8>(dummy_shape, dummy_tensor);
+  TF_ASSERT_OK(RunOpKernel());
+  // first calculate C = A * B,
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op_test.cc]:
```suggestion
  // First calculate C = A * B,
```

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc [hunk 10]
```diff
@@ -0,0 +1,464 @@
...
+          .Finalize(node_def()));
+  TF_ASSERT_OK(InitOp());
+  // The A matrix is:
+  // |  -1 |  -5 |  -9 |
+  // |  -2 |  -6 | -10 |
+  // |  -3 |  -7 | -11 |
+  // |  -4 |  -8 | -12 |
+  // The input array only contains unsigned bytes, so we specify the actual
+  // quantized values as Au8 = (Af32 - Min(A𝑓32)*Qa, where
+  // Qa = 255/(Max(A𝑓32)–Min(A𝑓32)). For eg. that means -1 is represented
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op_test.cc]:
```suggestion
  // Qa = 255/(Max(A𝑓32)–Min(A𝑓32)). For example, -1 is represented
```

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc [hunk 11]
```diff
@@ -0,0 +1,464 @@
...
+  const Tensor& mkl_shape_tensor = *GetOutput(3);
+  ConvMklToTF conv_comp;
+  Tensor output_quantized;
+  conv_comp.ConvertMKL2TF<qint32>(DT_QINT32, output, mkl_shape_tensor,
+                                  output_quantized);
+
+  test::ExpectTensorEqual<qint32>(expected, output_quantized);
+}
+
+// simple test for Matrix multiplication with Bias, Relu and
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op_test.cc]:
```suggestion
// Simple test for Matrix multiplication with Bias, Relu and
```

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc [hunk 12]
```diff
@@ -0,0 +1,464 @@
...
+  // | 15 | 16 | 17 | 18 |
+  AddInputFromArray<qint8>(TensorShape({3, 4}),
+                           {7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18});
+  AddInputFromArray<qint32>(TensorShape({4}), {10, -20, 30, -40});
+  AddInputFromArray<float>(TensorShape({1}), {0});
+  AddInputFromArray<float>(TensorShape({1}), {255.0f});
+  AddInputFromArray<float>(TensorShape({1}), {-127.0f});
+  AddInputFromArray<float>(TensorShape({1}), {127.0f});
+  AddInputFromArray<float>(TensorShape({1}), {0});
+  AddInputFromArray<float>(TensorShape({1}), {256.0f});
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op_test.cc]: Just curious, is 256 (not 255) intentional?

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc [hunk 13]
```diff
@@ -0,0 +1,464 @@
...
+  ConvMklToTF conv_comp;
+  Tensor output_quantized;
+  conv_comp.ConvertMKL2TF<quint8>(DT_QUINT8, output, mkl_shape_tensor,
+                                  output_quantized);
+
+  test::ExpectTensorEqual<quint8>(expected, output_quantized);
+}
+
+}  // namespace tensorflow
+#endif
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op_test.cc]: Please add a newline before this line. Also,
```suggestion
#endif  // INTEL_MKL
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR! I appreciate the detailed comments. I have minor formatting comments.

**Comment by mdfaijul** [comment]
@penpornk Thank you very much for the comments. Please check the latest commit.

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc [hunk 14]
```diff
@@ -0,0 +1,464 @@
...
+          .Attr("_kernel", "QuantizedMklOp")
+          .Finalize(node_def()));
+  TF_ASSERT_OK(InitOp());
+  // The A matrix is:
+  // |  -1 |  -5 |  -9 |
+  // |  -2 |  -6 | -10 |
+  // |  -3 |  -7 | -11 |
+  // |  -4 |  -8 | -12 |
+  // The input array only contains unsigned bytes, so we specify the actual
+  // quantized values as Au8 = (Af32 - Min(A𝑓32) * Qa, where
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_qmatmul_op_test.cc]: Missing one closing parenthesis.
```suggestion
  // quantized values as Au8 = (Af32 - Min(A𝑓32)) * Qa, where
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for your quick fixes! One more change and this should be good to go too.

File: tensorflow/core/kernels/mkl_qmatmul_op_test.cc (hunk 14 above)
Comment (mdfaijul) [reply tensorflow/core/kernels/mkl_qmatmul_op_test.cc]: @penpornk Thank you very much. Please check the latest commit.

**Review by penpornk — COMMENTED** [review]
Thank you! I'll approve this when parts 2 and 3 are merged. (If I approve it now the tests will just fail.)

**Review by penpornk — APPROVED** [review]
Approving this since part 1 (https://github.com/tensorflow/tensorflow/pull/26909) has been merged.

**Comment by penpornk** [comment]
@mdfaijul Some of the failed tests are from the missing `mkl_qmatmul_op` target. (My bad, I thought the build is also guarded by `is_mkl`. I'll look into that after all three parts are merged.) We still have some time. I'll wait until part 2 is merged and rerun the tests then.

**Pull request merged**